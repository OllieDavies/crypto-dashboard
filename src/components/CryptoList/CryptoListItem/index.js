import React from "react";
import { Link } from "react-router-dom";

import ChangeIndicator from "../../../components/ChangeIndicator";
import styles from "./styles.module.css";

const CryptoListItem = ({ id, name, symbol, price, change }) => (
  <Link to={`./${id}`} className={styles.wrapper} data-testid="list-item-link">
    <div className={styles.container}>
      <div className={styles.nameContainer}>
        <h4 className={styles.name}>{name}</h4>
        <p className={styles.price}>
          {price} {symbol}
        </p>
      </div>
      <div className={styles.changeContainer}>
        <ChangeIndicator change={change} />
      </div>
    </div>
  </Link>
);

export default CryptoListItem;
