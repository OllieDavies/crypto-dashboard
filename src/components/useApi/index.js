import { useState, useCallback } from "react";
import { get } from "axios";

export default function useApi(path) {
  const [isLoading, setLoading] = useState(true);
  const [response, setResponse] = useState(undefined);
  const [error, setError] = useState(undefined);

  const fetch = useCallback(() => {
    setLoading(true);
    get(path)
      .then(setResponse)
      .catch(setError)
      .finally(() => setLoading(false));
  }, [path]);

  return {
    isLoading,
    response,
    error,
    fetch,
  };
}
