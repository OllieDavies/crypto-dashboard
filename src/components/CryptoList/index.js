import React, { useEffect } from "react";

import useApi from "../useApi";
import CryptoListItem from "./CryptoListItem";

const CryptoList = ({ order }) => {
  const { isLoading, response, error, fetch } = useApi(
    `https://api.coinranking.com/v1/public/coins?base=GBP&timePeriod=7d&sort=price&order=${order}`
  );

  // Fetch on mount and start a timer to poll
  useEffect(() => {
    fetch();
    const timer = setInterval(fetch, 3000);
    return () => clearInterval(timer);
  }, [fetch]);

  if (isLoading === true && response === undefined) return <h1>Loading...</h1>;
  if (error) return <h1>An error occured</h1>;

  return (
    <>
      {response.data.data.coins.map((coin) => (
        <CryptoListItem key={coin.id} {...coin} />
      ))}
    </>
  );
};

export default CryptoList;
