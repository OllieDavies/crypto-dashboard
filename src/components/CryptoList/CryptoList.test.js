import React from "react";
import { render } from "@testing-library/react";
import useApi from "../useApi";

import CryptoList from ".";

jest.mock("../useApi");
jest.mock("./CryptoListItem", () => () => (
  <div data-testid="listitem">List Item</div>
));

describe("<CryptoList />", () => {
  it("matches snapshot when first mounted", () => {
    useApi.mockReturnValue({
      isLoading: true,
      fetch: jest.fn(),
    });

    const { container } = render(<CryptoList />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches snapshot when an error is returned", () => {
    useApi.mockReturnValue({
      isLoading: false,
      fetch: jest.fn(),
      error: new Error("Error"),
    });

    const { container } = render(<CryptoList />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("displays the correct number of CryproListItem's when data is returned", () => {
    useApi.mockReturnValue({
      isLoading: false,
      fetch: jest.fn(),
      response: {
        data: {
          data: {
            coins: [
              {
                id: 1,
                name: "Bitcoin",
                symbol: "BTC",
                price: "7500",
                change: "0.3",
              },
              {
                id: 2,
                name: "Litecoin",
                symbol: "LTC",
                price: "120",
                change: "20",
              },
            ],
          },
        },
      },
    });

    const { queryAllByTestId } = render(<CryptoList />);
    expect(queryAllByTestId("listitem").length).toBe(2);
  });
});
