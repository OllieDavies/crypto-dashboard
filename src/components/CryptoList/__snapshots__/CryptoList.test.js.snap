// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CryptoList /> matches snapshot when an error is returned 1`] = `
<h1>
  An error occured
</h1>
`;

exports[`<CryptoList /> matches snapshot when first mounted 1`] = `
<h1>
  Loading...
</h1>
`;
